### PR TITLE
Bash completions for etcdctl

### PR DESCRIPTION
**Pull request by eparis**
This does 2 things.  First is ports from codegansta/cli to spf13/cobra.

Second it uses spf13/cobra to generate intelligent bash completions for etcdctl.  The are generated by running

`etcdctl genbash --outfile=yourfile` and then sourcing the file `. yourfile` (or you could put the output in `/etc/bash_completion.d/` or `/usr/share/bash-completion/completions/` and have it automatically loaded.)

These completions are pretty cool because on my machine you can do obvious simple things:

```
# etcdctl [tab][tab]
backup          exec-watch      get             import          member          mkdir           rmdir           setdir          updatedir       
cluster-health  genbash         help            ls              mk              rm              set             update          watch

# etcdctl member [tab][tab]
add     list    remove  

```

But it also will show you data on 127.0.0.1!

```
# etcdctl ls /[tab][tab]
/coreos.com  /kube.local  /registry
```

```
# etcdctl ls /r[tab][tab]

completes

# etcdctl ls /registry/
/registry/controllers      /registry/minions          /registry/pods             /registry/serviceaccounts  
/registry/events           /registry/namespaces       /registry/ranges           /registry/services
```

It is NOT smart enough to use things you already types on your command line, so --peers etc aren't used trying to show the completions.  Only loopback.

### CODE REVIEW

**Comment by eparis** [comment]
This is a WIP because I didn't immediately know how to handle `--hidden` to `etcdctl import` so I just disabled it. If people like this change, I'll figure that out. If not I thought, nah, forget it, yo, homes to Bel Air.

**Comment by eparis** [comment]
Also I realize I'll have to handle godeps 'the coreos' way. But I already opened another PR (#2893) to try to fix etcd's Godeps before I try to add more!

I could include that fix here, and give the complete solution, but review would be harder...

**Comment by eparis** [comment]
in case you want to try the results without fighting through building this code:

https://gist.github.com/eparis/2380a2daa7ff6c0b52a6

**Comment by eparis** [comment]
updated, now with godeps, so building is easier/possible

File: etcdctl/command/import_snap_command.go [hunk 1]
```diff
@@ -85,14 +87,14 @@ func handleImportSnap(c *cli.Context) {
...
-		n += copyKeys(allh.Node, setc)
-	}
+	//hiddens := c.StringSlice("hidden")
+	//for _, h := range hiddens {
+	//allh, err := st.Get(h, true, true)
+	//if err != nil {
+	//handleError(ErrorFromEtcd, err)
+	//}
+	//n += copyKeys(allh.Node, setc)
+	//}
```
Comment (eparis) [inline etcdctl/command/import_snap_command.go]: THIS MUST BE FIXED BEFORE MERGE

**Comment by xiang90** [comment]
@eparis I used cobra in my side project. 

I found it is a little bit annoying (maybe it is configurable or my memory is wrong) is that I need to use "=" to separate the flag the value if I use the long flag (--flag). 

If I use short flag, it behavior differently and unexpectedly.  For example. I have a short flag "-d". If I use "-data", it will set -d flag to ata...

File: etcdctl/command/import_snap_command.go (hunk 1 above)
Comment (xiang90) [inline etcdctl/command/import_snap_command.go]: yea... another I noticed is that cobra does not support slice.

**Comment by eparis** [comment]
https://github.com/spf13/pflag/pull/20 has a fix to the --flag "bob" not working.

I can probably poke steve to see what he thinks about me fixing it (add adding optional args at the same time)

-data is 'expected' if -d isn't a boolean...   i don't see a way 'fix' that....

File: etcdctl/command/import_snap_command.go (hunk 1 above)
Comment (eparis) [inline etcdctl/command/import_snap_command.go]: https://github.com/eparis/pflag/tree/SliceStrings

Has support for string slices. I can define it in the etcdctl code, but that'd suck, so i'd consider this a prereq before I remove the WIP....

**Comment by xiang90** [comment]
@eparis My argument is short and long behavior differently...

Look this, I have a flag (data, d). The app simply prints out the flag.

```
./myapp --dataddd
Error: unknown flag: --dataddd
Run 'myapp help' for usage.
```

```
./myapp -dataddd
ataddd
```

**Comment by eparis** [comment]
@xiang90 that's called posix flag parsing. I get it, but that's the way unix does it. Which is why I don't think it can be 'fixed'

```
# 11 entries in dir
$ ls | wc -l
11

# ignore one of them
$ ls --ignore bob | wc -l
10

# can't just mash them together with long args
$ ls --ignorebob | wc -l
ls: unrecognized option '--ignorebob'
Try 'ls --help' for more information.
0

# space wtih short args works
$ ls -I bob | wc -l
10

# and with short args you can just mash them together
$ ls -Ibob | wc -l
10
```

**Comment by xiang90** [comment]
@eparis OK. So I will be happy to take a further look at this when:

1. slice flag is resolved
2. --foo "Bob" works.

Sounds good?

**Comment by eparis** [comment]
did you play with the completion?  was it useful enough for me to keep pushing?   (at the least --foo "bob" needs fixed no matter what!)

**Comment by xiang90** [comment]
@eparis Yea. I play with it. I think it is quite useful. But to be honest, I use etcd client more often than ctl as a developer. 

@philips Probably you have an opinion on the completions?

**Comment by eparis** [comment]
All comments should be addressed and this should be g2g....

**Comment by xiang90** [comment]
@eparis This is too late for 2.1. Can you send a pull request to dev 2.2 branch?

**Comment by gyuho** [comment]
@eparis Any updates on this? 

BTW, we have migrated to `spf13/cobra` for V3.

Please check that out if you have time.

**Comment by xiang90** [comment]
Closing this since it is inactive.

**Comment by tpo** [comment]
@eparis a pitty this hasn't gone in. I'd love to have etcdctl bash completion. Any chance you pick this up again and try to push this into etcd?

**Pull request closed without merging**